Autopilot/AttitudeManager/OrientationControl.c

```diff
@@ -1,146 +1,140 @@
     }
     
-    //Derivative Calculations
-    int dValue = error - lastError[HEADING];
-    lastError[HEADING] = error;
+    pid->lastTime = now;
 
-    int controlSignal = (int)(HEADING_ROLL_SCALE_FACTOR * ((dValue * kd_gain[HEADING]) + (error * kp_gain[HEADING]) + (sum_gain[HEADING] * ki_gain[HEADING])));
-    return controlSignal;
-}
-int controlSignalAngles(float setpoint, float output, unsigned char type, float SERVO_SCALE_FACTOR_ANGLES) { // function to find output based on gyro acceleration and PWM input
+    output += pid->Kp * error; // Proportional control
+    
+    if (delta_msec > 0) { // only compute time-sensitive control if time has elapsed
+        float dTime = delta_msec / 1000.0f; // elapsed time in seconds
+        
+        if (fabsf(pid->Ki) > 0) { // Integral control
+            pid->integral += (pid->Ki * error) * dTime;
+            
+            if (pid->integral < -pid->imax) { // ensure integral stays manageable
+                pid->integral = -pid->imax;
+            } else if (pid->integral > pid->imax) {
+                pid->integral = pid->imax;
+            }
+            output += pid->integral;
+        }
 
-    float error = setpoint - output;
-    if (integralFreeze == 0){
-        sum_gain[type] += error;
+        if (fabsf(pid->Kd) > 0) { // Derivative control
+            float derivative = (error - pid->lastErr) / dTime;
+            derivative = derivative * FILTER + pid->lastDer * (1-FILTER); // reduce jitter in derivative by averaging
+            pid->lastErr = error;
+            pid->lastDer = derivative;
+            output += pid->Kd * derivative;
+        }
     }
-
-    int controlSignal = (int)(SERVO_SCALE_FACTOR_ANGLES * (error * kp_gain[type] + (sum_gain[type] * ki_gain[type])));
-    return controlSignal;
-}
-int controlSignal(float setpoint, float output, unsigned char type) { // function to find output based on gyro acceleration and PWM input
-    int controlSignal = (int)(SERVO_SCALE_FACTOR * (setpoint - output * kd_gain[type]));
-    return controlSignal;
+    
+    return output * scale;
 }
 
-void freezeIntegral() {
-    integralFreeze = 1;
-}
+/*
+ * Orientation-controller functions begin below here
+ */
 
-void unfreezeIntegral() {
-    integralFreeze = 0;
+void orientationInit() {
+    uint8_t i;
+    for (i = 0; i < PID_CHANNELS; i++) {
+        initPID(&pids[i], init_kp[i], init_ki[i], init_kd[i], 1000);
+    }
 }
 
-void setIntegralSum(unsigned char YPRH, float value) {
-    sum_gain[YPRH] = value;
-}
-float getIntegralSum(unsigned char YPRH){
-    return sum_gain[YPRH];
+PID_val* getPID(PID_channel channel) {
+    return &pids[channel];
 }
 
-float getGain(unsigned char YPRH, unsigned char type){
-    if (type == GAIN_KD){
-        return kd_gain[YPRH];
-    }
-    else if (type == GAIN_KP){
-        return kp_gain[YPRH];
-    }
-    else if (type == GAIN_KI){
-        return ki_gain[YPRH];
+
+float getGain(PID_channel channel, PID_gain type){
```

Comment:
not sure if i like the inconsistent accessor interface. Some of the functions here take a pointer to a PID_val and this one takes in a channel?